x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -72,50 +82,140 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
+                        List<IndexShardStats> indexShardStatsList = nodeStatsMap.get(nodeId).getIndices().
+                            getShardStats(indexMetadata.getIndex());
+                        long shardsOnCurrentNodeStorageBytes = indexShardStatsList.stream().mapToLong(indexShardStats ->
+                            Arrays.stream(indexShardStats.getShards()).mapToLong(shardStats ->
+                                shardStats.getStats().getStore() == null ? 0 :
+                                    shardStats.getStats().getStore().getSizeInBytes()).sum()).sum();
+                        if (shardsOnCurrentNodeStorageBytes != 0) {
+                            nodeShardsStorageBytes.put(nodeId, shardsOnCurrentNodeStorageBytes);
+                        }
+                    }
+                }
+                long indexPrimaryShardsStorageBytes = nodeShardsStorageBytes.values().stream().mapToLong(Long::longValue).sum();
+                if (indexMetadata.getNumberOfReplicas() != 0) {
+                    indexPrimaryShardsStorageBytes /= indexMetadata.getNumberOfReplicas();
+                }
+                for (String nodeId : dataNodeIds) {
+                    if (nodeStatsMap.get(nodeId) != null) {
+                        FsInfo.Path totalFsInfo = nodeStatsMap.get(nodeId).getFs().getTotal();
+                        long nodeTotalBytes = totalFsInfo.getTotal().getBytes();
+                        long nodeAvailableBytes = totalFsInfo.getAvailable().getBytes();
+                        long freeBytesThresholdLow;
+                        if (diskThresholdSettings.getFreeDiskThresholdLow() != 0) {
+                            freeBytesThresholdLow = (long) Math.ceil(nodeTotalBytes *
+                                diskThresholdSettings.getFreeDiskThresholdLow() * 0.01);
+                        } else {
+                            freeBytesThresholdLow = diskThresholdSettings.getFreeBytesThresholdLow().getBytes();
+                        }
+                        // we should make sure that allocating one copy of the index's primary shards to that node doesn't put the node
+                        // above the low watermark, if not, the new shrunken index may not be initialized successfully
+                        long shardsOnCurrentNodeStorageBytes = 0;
+                        if (nodeShardsStorageBytes.containsKey(nodeId)) {
+                            shardsOnCurrentNodeStorageBytes = nodeShardsStorageBytes.get(nodeId);
+                        }
+                        if (nodeAvailableBytes > freeBytesThresholdLow + indexPrimaryShardsStorageBytes -
+                            shardsOnCurrentNodeStorageBytes) {
+                            validNodeIds.add(nodeId);
+                        }
+                    }
+                }
+
+                if (validNodeIds.size() == 0) {
+                    logger.debug("no nodes have enough disk space to hold one copy of the index [{}] onto prior to shrink ", indexName);
```

Comment:
I believe this error message is incorrect as we're including the size of the (future) shrunken index in the math as well? We should reflect that in the message, unless I'm misreading this.

Reply:
No, we only consider the size of the source index here.